subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/capabilities/CapabilitiesUseCasesIntegrationTest.groovy

```diff
@@ -0,0 +1,415 @@
+     * for "replacedBy" rules when a module has been relocated, and that the publisher uses Gradle. There's the
+     * ability to declare, in a newer version of the module, that it actually provides the same capability as an
+     * older version published at different coordinates.
+     *
+     * This test also makes sure that the order in which dependencies are seen in the graph do not matter.
+     */
+    @RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+    )
+    @Unroll
+    def "published module can declare relocation (first in graph = #first, second in graph = #second)"() {
+        given:
+        repository {
+            'asm:asm:3.0' {
+                variant('runtime') {
+                    capability('asm', 'asm', '3.0')
+                }
+            }
+            'org.ow2.asm:asm:4.0' {
+                variant('runtime') {
+                    capability('asm', 'asm', '4.0') // upgrades the asm capability
+                    capability('org.ow2.asm', 'asm', '4.0') // self capability
+                }
+            }
+        }
+
+        buildFile << """
+            dependencies {
+               conf "$first"
+               conf "$second"
+            }
+        """
+
+        when:
+        repositoryInteractions {
+            'asm:asm:3.0' {
+                expectGetMetadata()
+            }
+            'org.ow2.asm:asm:4.0' {
+                expectResolve()
+            }
+        }
+        run ":checkDeps"
+
+        then:
+        resolve.expectGraph {
+            root(":", ":test:") {
+                edge('asm:asm:3.0', 'org.ow2.asm:asm:4.0')
+                    .byConflictResolution()
+                module('org.ow2.asm:asm:4.0')
+            }
+        }
+
+        where:
+        first << ['asm:asm:3.0', 'org.ow2.asm:asm:4.0']
+        second << ['org.ow2.asm:asm:4.0', 'asm:asm:3.0']
+    }
+
+
+    /**
+     * This test illustrates that published modules can declare capabilities, which are then discovered
+     * as we visit the graph. But using a module substitution rule, we can fix the problem.
+     */
+
+    @RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+    )
+    def "can express preference for capabilities declared in published modules"() {
+        given:
+        repository {
+            'org:testA:1.0' {
+                variant('runtime') {
+                    capability('cap')
+                }
+            }
+            'org:testB:1.0' {
+                variant('runtime') {
+                    capability('cap')
+                }
+            }
+        }
+
+        buildFile << """
+            dependencies {
+                conf 'org:testA:1.0'
+                conf 'org:testB:1.0'
+            }
+            
+            // fix the conflict between modules providing the same capability
```

Comment:
I'd prefer fail first, then fix. (Assuming the repository interactions don't make this near-impossible :) )

Reply:
done